Source/MQTTnet/Server/MqttClientSession.cs

```diff
@@ -153,6 +182,8 @@ public void Stop(MqttClientDisconnectType type)
                 }
 
                 _willMessage = null;
+
+                _workerTask?.GetAwaiter().GetResult();
```

Comment:
Is there a way to scip the wait if the calling thread is the thread which must be awaited? Can we check this by the thread ID or task ID etc.?

Reply:
Ah yes I remember such error. I was also fighting them and tried to avoid them with the wait calls. I will try to find another solution.

Reply:
> When a client is doing a disconnect, the _adapter is disconnected and disposed in Cleanup triggered by the cancellationToke.Register(...).

for me it sounds like a similar issue like in the client with the two disconnects overlapping so i guess we could also solve this with a tcs. for me it is more important to add tests for that case becasue I assume everything is fine if the tests are green. 

> I don't think the code is necessary at all
on my machine it was otherwise the server socket is closed before the disconnect message is send to all clients

Reply:
> ... add tests for that case becasue I assume everything is fine if the tests are green.

If you are interested, so here are my test, with a very temporary name, and some temporary code (the logging).
https://github.com/jenscski/MQTTnet/blob/45c125b86d675cc45bb88803100a91b4f411d495/Tests/MQTTnet.Core.Tests/MqttServerTests.cs#L265-L303

Reply:
@jenscski thx for the testing method i added it to my pr and fixed the clean disconnect. not very elegant but hey it works... canyou please verify if you have the time?